cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -35,23 +37,59 @@ namespace cudf {
 namespace groupby {
 namespace detail {
 namespace {
+
+/**
+ * @brief Functor to compare two rows of a table in given permutation order
+ * This is useful to identify unique elements in a sorted order table, when the permutation order is
+ * the sorted order of the table.
+ *
+ */
+template <typename Iterator>
+struct permuted_comparator {
+  /**
+   * @brief comparator object which compares two rows of the table in given permutation order
+   *
+   * @param device_table Device table to compare
+   * @param permute permutation order, integer type column.
+   * @param has_nulls whether the table has nulls
+   */
+  permuted_comparator(table_device_view device_table, Iterator const permute_begin, bool has_nulls)
+    : comparator(nullate::DYNAMIC{has_nulls}, device_table, device_table, null_equality::EQUAL),
+      permute(permute_begin)
+  {
+  }
+  __device__ bool operator()(size_type index1, size_type index2) const
+  {
+    return comparator(permute[index1], permute[index2]);
+  };
+
+ private:
+  row_equality_comparator<nullate::DYNAMIC> comparator;
+  Iterator const permute;
```

Comment:
"permute" is a verb, making it seem like this is a function, but really it's an iterator containing the permutation so we should use a noun for the name
```suggestion
  Iterator const permutation;
```